File: packages/cli/src/commands/doctor.ts
import { existsSync, statSync, readdirSync, readFileSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { spawnSync } from "node:child_process";
import { Resolver, lookup } from "node:dns/promises";
import { PROTOCOL_VERSION } from "@yuanio/shared";

const YUANIO_DIR = `${process.env.HOME || process.env.USERPROFILE}/.yuanio`;
const DEFAULT_CONTROL_SERVER = "http://localhost:3000";

interface CheckResult {
  label: string;
  ok: boolean;
  detail: string;
}

interface DoctorOptions {
  controlServerUrl: string;
  publicServerUrl?: string;
}

// ── 检查项 ──

function checkAgent(name: string, cmd: string): CheckResult {
  try {
    const result = spawnSync(cmd, ["--version"], {
      timeout: 5000,
      stdio: ["ignore", "pipe", "pipe"],
      shell: true,
    });
    if (result.status === 0) {
      const version = result.stdout?.toString().trim().split("\n")[0] || "unknown";
      return { label: `${name} CLI`, ok: true, detail: version };
    }
    return { label: `${name} CLI`, ok: false, detail: "not found" };
  } catch {
    return { label: `${name} CLI`, ok: false, detail: "not found" };
  }
}

async function checkUrlReachability(label: string, url: string): Promise<CheckResult> {
  try {
    const res = await fetch(`${url}/health`, { signal: AbortSignal.timeout(5000) });
    if (res.ok) {
      return { label, ok: true, detail: `connected (${url})` };
    }
    return { label, ok: false, detail: `HTTP ${res.status} (${url})` };
  } catch (e: any) {
    return { label, ok: false, detail: `${e?.message || "unreachable"} (${url})` };
  }
}

async function checkProtocolCompatibility(controlServerUrl: string): Promise<CheckResult> {
  try {
    const res = await fetch(`${controlServerUrl}/health`, { signal: AbortSignal.timeout(5000) });
    if (!res.ok) {
      return { label: "Protocol", ok: false, detail: `health HTTP ${res.status}` };
    }
    const json = await res.json() as { protocolVersion?: string };
    const serverVersion = json.protocolVersion;
    if (!serverVersion) {
      return { label: "Protocol", ok: true, detail: "legacy relay (version unknown)" };
    }
    const sameMajor = serverVersion.split(".")[0] === PROTOCOL_VERSION.split(".")[0];
    if (!sameMajor) {
      return {
        label: "Protocol",
        ok: false,
        detail: `major mismatch cli=${PROTOCOL_VERSION} relay=${serverVersion}`,
      };
    }
    return { label: "Protocol", ok: true, detail: `compatible cli=${PROTOCOL_VERSION} relay=${serverVersion}` };
  } catch (e: any) {
    return { label: "Protocol", ok: false, detail: e?.message || "check failed" };
  }
}

function checkKeys(): CheckResult {
  const keystoreFile = join(YUANIO_DIR, "keys.json");
  if (!existsSync(keystoreFile)) {
    return { label: "Keys", ok: false, detail: "keystore not found" };
  }
  try {
    const data = JSON.parse(readFileSync(keystoreFile, "utf-8"));
    if (data.sessionToken && data.secretKey && data.peerPublicKey) {
      return { label: "Keys", ok: true, detail: "valid" };
    }
    return { label: "Keys", ok: false, detail: "incomplete keystore" };
  } catch {
    return { label: "Keys", ok: false, detail: "corrupted keystore" };
  }
}

function checkDaemon(): CheckResult {
  const stateFile = join(YUANIO_DIR, "daemon.json");
  if (!existsSync(stateFile)) {
    return { label: "Daemon", ok: false, detail: "not running" };
  }
  try {
    const state = JSON.parse(readFileSync(stateFile, "utf-8"));
    process.kill(state.pid, 0);
    return { label: "Daemon", ok: true, detail: `PID=${state.pid} port=${state.port}` };
  } catch {
    return { label: "Daemon", ok: false, detail: "not running (stale state)" };
  }
}

function checkDisk(): CheckResult {
  if (!existsSync(YUANIO_DIR)) {
    return { label: "Disk", ok: true, detail: "0 bytes" };
  }
  let totalSize = 0;
  try {
    const walk = (dir: string) => {
      for (const entry of readdirSync(dir, { withFileTypes: true })) {
        const full = join(dir, entry.name);
        if (entry.isDirectory()) walk(full);
        else totalSize += statSync(full).size;
      }
    };
    walk(YUANIO_DIR);
  } catch {}
  const mb = (totalSize / (1024 * 1024)).toFixed(1);
  return { label: "Disk", ok: true, detail: `${mb}MB used` };
}

function checkProxyEnv(): CheckResult {
  const keys = ["HTTP_PROXY", "HTTPS_PROXY", "http_proxy", "https_proxy"];
  const active = keys
    .map((key) => ({ key, value: process.env[key] }))
    .filter((item) => typeof item.value === "string" && item.value.trim().length > 0);
  if (active.length === 0) {
    return { label: "Proxy", ok: true, detail: "not set" };
  }
  const values = active.map((item) => `${item.key}=${item.value}`).join("; ");
  return { label: "Proxy", ok: false, detail: values };
}

function checkCloudflaredBinary(): CheckResult {
  try {
    const result = spawnSync("cloudflared", ["--version"], {
      timeout: 5000,
      stdio: ["ignore", "pipe", "pipe"],
      shell: true,
    });
    if (result.status === 0) {
      const version = result.stdout?.toString().trim().split("\n")[0] || "unknown";
      return { label: "Cloudflared", ok: true, detail: version };
    }
    return { label: "Cloudflared", ok: false, detail: "not found" };
  } catch {
    return { label: "Cloudflared", ok: false, detail: "not found" };
  }
}

function checkCloudflaredServiceOnWindows(): CheckResult | null {
  if (process.platform !== "win32") return null;
  try {
    const result = spawnSync("sc", ["query", "cloudflared"], {
      timeout: 5000,
      stdio: ["ignore", "pipe", "pipe"],
      shell: true,
    });
    const output = `${result.stdout?.toString() || ""}\n${result.stderr?.toString() || ""}`;
    if (result.status !== 0 && /does not exist|1060/i.test(output)) {
      return { label: "CF Service", ok: false, detail: "not installed" };
    }
    if (/RUNNING/i.test(output)) {
      return { label: "CF Service", ok: true, detail: "running" };
    }
    if (/STOPPED/i.test(output)) {
      return { label: "CF Service", ok: false, detail: "stopped" };
    }
    return { label: "CF Service", ok: false, detail: "unknown state" };
  } catch {
    return { label: "CF Service", ok: false, detail: "query failed" };
  }
}

async function resolveHostWith(resolver: Resolver, host: string): Promise<Set<string>> {
  const values = new Set<string>();
  try {
    for (const ip of await resolver.resolve4(host)) values.add(ip);
  } catch {}
  try {
    for (const ip of await resolver.resolve6(host)) values.add(ip);
  } catch {}
  return values;
}

async function resolveHostByLookup(host: string): Promise<Set<string>> {
  const values = new Set<string>();
  try {
    const records = await lookup(host, { all: true });
    for (const record of records) {
      if (record.address) values.add(record.address);
    }
  } catch {}
  return values;
}

function isLikelyIp(host: string): boolean {
  return /^\d{1,3}(\.\d{1,3}){3}$/.test(host) || host.includes(":");
}

async function checkPublicDns(publicServerUrl?: string): Promise<CheckResult | null> {
  if (!publicServerUrl) return null;
  let host: string;
  try {
    host = new URL(publicServerUrl).hostname;
  } catch {
    return { label: "DNS", ok: false, detail: "public server URL invalid" };
  }

  if (!host || host === "localhost" || isLikelyIp(host)) {
    return { label: "DNS", ok: true, detail: "skipped (localhost/ip)" };
  }

  const systemResolver = new Resolver();
  const publicResolver = new Resolver();
  publicResolver.setServers(["1.1.1.1", "8.8.8.8"]);

  const [systemIps, publicIps] = await Promise.all([
    resolveHostWith(systemResolver, host),
    resolveHostWith(publicResolver, host),
  ]);
  if (systemIps.size === 0) {
    const lookupIps = await resolveHostByLookup(host);
    for (const ip of lookupIps) systemIps.add(ip);
  }

  if (publicIps.size === 0) {
    return { label: "DNS", ok: false, detail: `public resolver failed (${host})` };
  }
  if (systemIps.size === 0) {
    return { label: "DNS", ok: false, detail: `system resolver failed (${host})` };
  }

  const overlap = [...systemIps].filter((ip) => publicIps.has(ip));
  const systemText = [...systemIps].slice(0, 3).join(",");
  const publicText = [...publicIps].slice(0, 3).join(",");

  if (overlap.length === 0) {
    return {
      label: "DNS",
      ok: false,
      detail: `mismatch system=[${systemText}] public=[${publicText}]`,
    };
  }
  return { label: "DNS", ok: true, detail: `aligned (${host})` };
}

// ── 僵尸进程清理 ──

function killRunawayProcesses(): number {
  let killed = 0;
  const stateFile = join(YUANIO_DIR, "daemon.json");
  if (!existsSync(stateFile)) return killed;

  try {
    const state = JSON.parse(readFileSync(stateFile, "utf-8"));
    try {
      process.kill(state.pid, 0);
    } catch {
      unlinkSync(stateFile);
      killed++;
    }
  } catch {}
  return killed;
}

// ── 主函数 ──

function normalizeDoctorOptions(input: string | Partial<DoctorOptions>): DoctorOptions {
  if (typeof input === "string") {
    return { controlServerUrl: input || DEFAULT_CONTROL_SERVER };
  }
  return {
    controlServerUrl: input.controlServerUrl || DEFAULT_CONTROL_SERVER,
    publicServerUrl: input.publicServerUrl,
  };
}

export async function runDoctor(
  options: string | Partial<DoctorOptions> = DEFAULT_CONTROL_SERVER,
): Promise<void> {
  const { controlServerUrl, publicServerUrl } = normalizeDoctorOptions(options);
  console.log("\n🔍 Yuanio Doctor\n");

  const results: CheckResult[] = [];

  results.push(checkAgent("Claude", "claude"));
  results.push(checkAgent("Codex", "codex"));
  results.push(checkAgent("Gemini", "gemini"));

  results.push(await checkUrlReachability("Relay(control)", controlServerUrl));
  results.push(await checkProtocolCompatibility(controlServerUrl));
  if (publicServerUrl && publicServerUrl !== controlServerUrl) {
    results.push(await checkUrlReachability("Relay(public)", publicServerUrl));
  }

  const dnsResult = await checkPublicDns(publicServerUrl);
  if (dnsResult) results.push(dnsResult);
  results.push(checkProxyEnv());

  results.push(checkKeys());
  results.push(checkDaemon());
  results.push(checkDisk());
  results.push(checkCloudflaredBinary());
  const serviceResult = checkCloudflaredServiceOnWindows();
  if (serviceResult) results.push(serviceResult);

  for (const r of results) {
    const icon = r.ok ? "✓" : "✗";
    console.log(`  ${icon} ${r.label}: ${r.detail}`);
  }

  const cleaned = killRunawayProcesses();
  if (cleaned > 0) {
    console.log(`\n  🧹 清理了 ${cleaned} 个僵尸进程`);
  }

  const failed = results.filter((r) => !r.ok);
  if (failed.length === 0) {
    console.log("\n  ✅ 所有检查通过\n");
  } else {
    console.log(`\n  ⚠️ ${failed.length} 项检查未通过\n`);
  }
}

File: yuanio-desktop/src-tauri/src/doctor.rs
use reqwest::blocking::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::{
    collections::HashSet,
    fs,
    net::IpAddr,
    path::{Path, PathBuf},
    process::Command,
};

use crate::{
    daemon_state::{read_state, remove_state_file, resolve_state_path},
    keystore::load_keys,
};

const PROTOCOL_VERSION: &str = "1.0.0";

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DoctorCheck {
    pub label: String,
    pub ok: bool,
    pub detail: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DoctorReport {
    pub checks: Vec<DoctorCheck>,
    pub failed: usize,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DoctorRequest {
    pub control_server_url: String,
    pub public_server_url: Option<String>,
}

fn http_client() -> Result<Client, String> {
    Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(|e| format!("创建 HTTP 客户端失败: {e}"))
}

fn check_command(name: &str, cmd: &str) -> DoctorCheck {
    match Command::new(cmd).arg("--version").output() {
        Ok(output) if output.status.success() => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let version = stdout.lines().next().unwrap_or("unknown").trim();
            DoctorCheck {
                label: format!("{name} CLI"),
                ok: true,
                detail: version.to_string(),
            }
        }
        _ => DoctorCheck {
            label: format!("{name} CLI"),
            ok: false,
            detail: "not found".to_string(),
        },
    }
}

fn check_proxy_env() -> DoctorCheck {
    let keys = ["HTTP_PROXY", "HTTPS_PROXY", "http_proxy", "https_proxy"];
    let mut active = Vec::new();
    for key in keys {
        if let Ok(value) = std::env::var(key) {
            if !value.trim().is_empty() {
                active.push(format!("{key}={value}"));
            }
        }
    }
    if active.is_empty() {
        DoctorCheck {
            label: "Proxy".to_string(),
            ok: true,
            detail: "not set".to_string(),
        }
    } else {
        DoctorCheck {
            label: "Proxy".to_string(),
            ok: false,
            detail: active.join("; "),
        }
    }
}

fn resolve_base_dir() -> PathBuf {
    if let Ok(path) = std::env::var("YUANIO_KEYSTORE_DIR") {
        return PathBuf::from(path);
    }
    let home = std::env::var("USERPROFILE")
        .or_else(|_| std::env::var("HOME"))
        .unwrap_or_else(|_| ".".to_string());
    Path::new(&home).join(".yuanio")
}

fn check_disk() -> DoctorCheck {
    let dir = resolve_base_dir();
    if !dir.exists() {
        return DoctorCheck {
            label: "Disk".to_string(),
            ok: true,
            detail: "0 bytes".to_string(),
        };
    }
    let mut total: u64 = 0;
    let mut stack = vec![dir];
    while let Some(path) = stack.pop() {
        if let Ok(entries) = fs::read_dir(&path) {
            for entry in entries.flatten() {
                let path = entry.path();
                if let Ok(meta) = entry.metadata() {
                    if meta.is_dir() {
                        stack.push(path);
                    } else {
                        total = total.saturating_add(meta.len());
                    }
                }
            }
        }
    }
    let mb = (total as f64) / (1024.0 * 1024.0);
    DoctorCheck {
        label: "Disk".to_string(),
        ok: true,
        detail: format!("{:.1}MB used", mb),
    }
}

fn check_keys() -> DoctorCheck {
    match load_keys() {
        Ok(Some(session)) => {
            let keys = &session.keys;
            let required = [
                ("sessionToken", keys.session_token.as_str()),
                ("secretKey", keys.private_key.as_str()),
                ("peerPublicKey", keys.peer_public_key.as_str()),
            ];
            let missing: Vec<String> = required
                .into_iter()
                .filter(|(_, value)| value.trim().is_empty())
                .map(|(name, _)| name.to_string())
                .collect();
            if missing.is_empty() {
                DoctorCheck {
                    label: "Keys".to_string(),
                    ok: true,
                    detail: "valid".to_string(),
                }
            } else {
                DoctorCheck {
                    label: "Keys".to_string(),
                    ok: false,
                    detail: format!("incomplete keystore ({})", missing.join(",")),
                }
            }
        }
        Ok(None) => DoctorCheck {
            label: "Keys".to_string(),
            ok: false,
            detail: "keystore not found".to_string(),
        },
        Err(_) => DoctorCheck {
            label: "Keys".to_string(),
            ok: false,
            detail: "corrupted keystore".to_string(),
        },
    }
}

fn check_daemon() -> (DoctorCheck, Option<DoctorCheck>) {
    let Some(state) = read_state() else {
        return (
            DoctorCheck {
                label: "Daemon".to_string(),
                ok: false,
                detail: "not running".to_string(),
            },
            None,
        );
    };
    if pid_exists(state.pid) {
        return (
            DoctorCheck {
                label: "Daemon".to_string(),
                ok: true,
                detail: format!("PID={} port={}", state.pid, state.port),
            },
            None,
        );
    }

    let cleanup = if resolve_state_path().exists() {
        remove_state_file().ok();
        Some(DoctorCheck {
            label: "Cleanup".to_string(),
            ok: true,
            detail: "stale daemon state removed".to_string(),
        })
    } else {
        None
    };

    (
        DoctorCheck {
            label: "Daemon".to_string(),
            ok: false,
            detail: "not running (stale state)".to_string(),
        },
        cleanup,
    )
}

fn pid_exists(pid: u32) -> bool {
    if std::env::consts::OS != "windows" {
        return true;
    }
    let output = Command::new("tasklist")
        .args(["/FI", &format!("PID eq {}", pid)])
        .output();
    let Ok(output) = output else {
        return false;
    };
    let text = format!(
        "{}\n{}",
        String::from_utf8_lossy(&output.stdout),
        String::from_utf8_lossy(&output.stderr)
    );
    if text.contains("No tasks are running") {
        return false;
    }
    text.contains(&pid.to_string())
}

fn check_cloudflared_binary() -> DoctorCheck {
    match Command::new("cloudflared").arg("--version").output() {
        Ok(output) if output.status.success() => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let version = stdout.lines().next().unwrap_or("unknown").trim();
            DoctorCheck {
                label: "Cloudflared".to_string(),
                ok: true,
                detail: version.to_string(),
            }
        }
        _ => DoctorCheck {
            label: "Cloudflared".to_string(),
            ok: false,
            detail: "not found".to_string(),
        },
    }
}

fn check_cloudflared_service_windows() -> Option<DoctorCheck> {
    if std::env::consts::OS != "windows" {
        return None;
    }
    match Command::new("sc").args(["query", "cloudflared"]).output() {
        Ok(output) => {
            let text = format!(
                "{}\n{}",
                String::from_utf8_lossy(&output.stdout),
                String::from_utf8_lossy(&output.stderr)
            );
            if text.contains("1060") || text.to_lowercase().contains("does not exist") {
                return Some(DoctorCheck {
                    label: "CF Service".to_string(),
                    ok: false,
                    detail: "not installed".to_string(),
                });
            }
            if text.contains("RUNNING") {
                return Some(DoctorCheck {
                    label: "CF Service".to_string(),
                    ok: true,
                    detail: "running".to_string(),
                });
            }
            if text.contains("STOPPED") {
                return Some(DoctorCheck {
                    label: "CF Service".to_string(),
                    ok: false,
                    detail: "stopped".to_string(),
                });
            }
            Some(DoctorCheck {
                label: "CF Service".to_string(),
                ok: false,
                detail: "unknown state".to_string(),
            })
        }
        Err(_) => Some(DoctorCheck {
            label: "CF Service".to_string(),
            ok: false,
            detail: "query failed".to_string(),
        }),
    }
}

fn is_likely_ip(host: &str) -> bool {
    host.parse::<IpAddr>().is_ok()
}

fn extract_ips(text: &str) -> HashSet<IpAddr> {
    let mut ips = HashSet::new();
    for token in text.split(|c: char| !c.is_ascii_hexdigit() && c != '.' && c != ':') {
        if token.is_empty() {
            continue;
        }
        if let Ok(ip) = token.parse::<IpAddr>() {
            ips.insert(ip);
        }
    }
    ips
}

fn resolve_host_by_nslookup(host: &str, server: Option<&str>) -> Result<HashSet<IpAddr>, String> {
    let mut cmd = Command::new("nslookup");
    cmd.arg(host);
    if let Some(server) = server {
        cmd.arg(server);
    }
    let output = cmd.output().map_err(|_| "nslookup not available".to_string())?;
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    let combined = format!("{stdout}\n{stderr}");
    let ips = extract_ips(&combined);
    Ok(ips)
}

fn check_public_dns(public_server_url: &str) -> Option<DoctorCheck> {
    let host = match reqwest::Url::parse(public_server_url) {
        Ok(url) => url.host_str().map(|value| value.to_string()),
        Err(_) => None,
    }?;

    if host.is_empty() || host == "localhost" || is_likely_ip(&host) {
        return Some(DoctorCheck {
            label: "DNS".to_string(),
            ok: true,
            detail: "skipped (localhost/ip)".to_string(),
        });
    }

    let system_ips = resolve_host_by_nslookup(&host, None).unwrap_or_default();
    let public_ips = resolve_host_by_nslookup(&host, Some("1.1.1.1")).unwrap_or_default();

    if public_ips.is_empty() {
        return Some(DoctorCheck {
            label: "DNS".to_string(),
            ok: false,
            detail: format!("public resolver failed ({host})"),
        });
    }
    if system_ips.is_empty() {
        return Some(DoctorCheck {
            label: "DNS".to_string(),
            ok: false,
            detail: format!("system resolver failed ({host})"),
        });
    }

    let overlap = system_ips.iter().filter(|ip| public_ips.contains(ip)).count();
    let system_text = system_ips.iter().take(3).map(|ip| ip.to_string()).collect::<Vec<_>>().join(",");
    let public_text = public_ips.iter().take(3).map(|ip| ip.to_string()).collect::<Vec<_>>().join(",");

    if overlap == 0 {
        return Some(DoctorCheck {
            label: "DNS".to_string(),
            ok: false,
            detail: format!("mismatch system=[{system_text}] public=[{public_text}]"),
        });
    }

    Some(DoctorCheck {
        label: "DNS".to_string(),
        ok: true,
        detail: format!("aligned ({host})"),
    })
}

fn check_url_reachability(label: &str, url: &str, client: &Client) -> DoctorCheck {
    let health = format!("{}/health", url.trim_end_matches('/'));
    match client.get(health).send() {
        Ok(res) if res.status().is_success() => DoctorCheck {
            label: label.to_string(),
            ok: true,
            detail: format!("connected ({url})"),
        },
        Ok(res) => DoctorCheck {
            label: label.to_string(),
            ok: false,
            detail: format!("HTTP {} ({url})", res.status()),
        },
        Err(err) => DoctorCheck {
            label: label.to_string(),
            ok: false,
            detail: format!("{} ({url})", err),
        },
    }
}

fn check_protocol(control_url: &str, client: &Client) -> DoctorCheck {
    let health = format!("{}/health", control_url.trim_end_matches('/'));
    let res = client.get(health).send();
    let Ok(res) = res else {
        return DoctorCheck {
            label: "Protocol".to_string(),
            ok: false,
            detail: "check failed".to_string(),
        };
    };
    if !res.status().is_success() {
        return DoctorCheck {
            label: "Protocol".to_string(),
            ok: false,
            detail: format!("health HTTP {}", res.status()),
        };
    }
    let payload: Value = res.json().unwrap_or(Value::Null);
    let server_version = payload
        .get("protocolVersion")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    if server_version.is_empty() {
        return DoctorCheck {
            label: "Protocol".to_string(),
            ok: true,
            detail: "legacy relay (version unknown)".to_string(),
        };
    }
    let client_major = PROTOCOL_VERSION.split('.').next().unwrap_or("");
    let server_major = server_version.split('.').next().unwrap_or("");
    if client_major != server_major {
        return DoctorCheck {
            label: "Protocol".to_string(),
            ok: false,
            detail: format!("major mismatch cli={} relay={}", PROTOCOL_VERSION, server_version),
        };
    }
    DoctorCheck {
        label: "Protocol".to_string(),
        ok: true,
        detail: format!("compatible cli={} relay={}", PROTOCOL_VERSION, server_version),
    }
}

#[tauri::command]
pub fn doctor_run(payload: DoctorRequest) -> Result<DoctorReport, String> {
    let control = payload.control_server_url.trim().to_string();
    if control.is_empty() {
        return Err("controlServerUrl 为空".to_string());
    }
    let client = http_client()?;
    let mut checks = vec![
        check_command("Claude", "claude"),
        check_command("Codex", "codex"),
        check_command("Gemini", "gemini"),
        check_url_reachability("Relay(control)", &control, &client),
        check_protocol(&control, &client),
    ];

    if let Some(public_url) = payload.public_server_url {
        let public = public_url.trim().to_string();
        if !public.is_empty() && public != control {
            checks.push(check_url_reachability("Relay(public)", &public, &client));
        }
        if let Some(dns_check) = check_public_dns(&public) {
            checks.push(dns_check);
        }
    }

    checks.push(check_proxy_env());
    checks.push(check_keys());
    let (daemon_check, cleanup_check) = check_daemon();
    checks.push(daemon_check);
    if let Some(cleanup) = cleanup_check {
        checks.push(cleanup);
    }
    checks.push(check_disk());
    checks.push(check_cloudflared_binary());
    if let Some(service) = check_cloudflared_service_windows() {
        checks.push(service);
    }

    let failed = checks.iter().filter(|item| !item.ok).count();
    Ok(DoctorReport { checks, failed })
}
